service/src/main/kotlin/fi/espoo/evaka/varda/VardaUpdateServiceV2.kt

```diff
@@ -396,30 +397,46 @@ fun sendPlacementToVarda(client: VardaClient, vardaDecisionId: Long, evakaServic
     return res.vardaPlacementId
 }
 
+private fun guardiansLiveInSameAddress(guardians: List<VardaGuardianWithId>): Boolean {
+    val validResidenceCodes = guardians.map { g -> g.asuinpaikantunnus }
+        .filterNotNull().filter { asuinpaikantunnus -> asuinpaikantunnus.length > 0 }
+    return validResidenceCodes.size == guardians.size && validResidenceCodes.toSet().size == 1
+}
+
 fun sendFeeDataToVarda(vardaClient: VardaClient, db: Database.Connection, newVardaServiceNeed: VardaServiceNeed, evakaServiceNeed: EvakaServiceNeedInfoForVarda, feeDataByServiceNeed: FeeDataByServiceNeed): List<Long> {
-    val feeResponseIds = feeDataByServiceNeed.feeDecisionIds.map { feeId ->
+    val guardians = getChildVardaGuardians(db, evakaServiceNeed.childId)
+    check(guardians.isNotEmpty()) { error("VardaUpdate: could not create fee data for ${evakaServiceNeed.id}: child has no guardians") }
+
+    val feeResponseIds: List<Long> = feeDataByServiceNeed.feeDecisionIds.mapNotNull { feeId ->
         try {
             logger.info { "VardaUpdate: trying to send fee data $feeId" }
             val decision = db.read { it.getFeeDecisionsByIds(listOf(feeId)) }.first()
-            val guardians = getFeeReceiverGuardians(db, evakaServiceNeed.childId, decision.headOfFamily.id)
-            sendFeeDecisionToVarda(
+
+            // If head of family is not a guardian, fee data is not supposed to be sent to varda (https://wiki.eduuni.fi/display/OPHPALV/Huoltajan+tiedot)
+            if (guardians.none { it.id == decision.headOfFamily.id }) {
+                logger.info { "VardaUpdate: won't send fee data for fee decision $feeId - head of family is not a guardian of ${evakaServiceNeed.childId}" }
+                null
+            } else sendFeeDecisionToVarda(
                 client = vardaClient,
                 decision = decision,
                 vardaChildId = newVardaServiceNeed.vardaChildId!!,
                 evakaServiceNeedInfoForVarda = evakaServiceNeed,
-                guardians = guardians
+                guardians = if (guardiansLiveInSameAddress(guardians)) guardians else guardians.filter { guardian -> guardian.id == decision.headOfFamily.id }
```

Comment:
This should part should probably be done to voucher value decisions as well. Could there be a way to share this whole code block with both fee and voucher value decisions so they don't get out of sync?